pkg/asset/quota/openstack/openstack.go

```
@@ -0,0 +1,119 @@
+	for _, gen := range []constraintGenerator{
+		controlPlane(config, ctrplConfigs),
+		compute(config, computeReplicas, computeConfigs),
+		others,
+	} {
+		ret = append(ret, gen()...)
+	}
+	return aggregate(ret)
+}
+
+func controlPlane(config *types.InstallConfig, machines []*openstackprovider.OpenstackProviderSpec) func() []quota.Constraint {
+	return func() []quota.Constraint {
+		var ret []quota.Constraint
+		for _, m := range machines {
+			ret = append(ret, machineFlavorCoresToQuota(m.Flavor, config), machineFlavorRAMToQuota(m.Flavor, config))
+		}
+		ret = append(ret, instanceConstraint(int64(len(machines))))
+		return ret
+	}
+}
+
+func compute(config *types.InstallConfig, replicas []int64, machines []*openstackprovider.OpenstackProviderSpec) func() []quota.Constraint {
+	return func() []quota.Constraint {
+		var ret []quota.Constraint
+		for idx, m := range machines {
+			coresConstraint := machineFlavorCoresToQuota(m.Flavor, config)
+			coresConstraint.Count = coresConstraint.Count * replicas[idx]
+			ramConstraint := machineFlavorRAMToQuota(m.Flavor, config)
+			ramConstraint.Count = ramConstraint.Count * replicas[idx]
+			ret = append(ret, instanceConstraint(int64(replicas[idx])), coresConstraint, ramConstraint)
+		}
+		return ret
+	}
+}
+
+func others() []quota.Constraint {
+	return []quota.Constraint{}
+}
+
+func aggregate(quotas []quota.Constraint) []quota.Constraint {
+	sort.SliceStable(quotas, func(i, j int) bool {
+		return quotas[i].Name < quotas[j].Name
+	})
+
+	i := 0
+	for j := 1; j < len(quotas); j++ {
+		if quotas[i].Name == quotas[j].Name {
+			quotas[i].Count += quotas[j].Count
+		} else {
+			i++
+			if i != j {
+				quotas[i] = quotas[j]
+			}
+		}
+	}
+	return quotas[:i+1]
+}
+
+//constraintGenerator generates a list of constraints.
+type constraintGenerator func() []quota.Constraint
+
+func machineFlavorCoresToQuota(f string, config *types.InstallConfig) quota.Constraint {
+	ci, _ := validation.GetCloudInfo(config)
```

We cannot ignore the error returned here.

---

I agree, a mistake on my side. I'll fix this.